Mathlib/Analysis/SpecialFunctions/ContinuousFunctionalCalculus/Rpow.lean

```diff
@@ -0,0 +1,347 @@
+-/
+
+import Mathlib.Analysis.SpecialFunctions.Pow.Real
+import Mathlib.Analysis.CstarAlgebra.ContinuousFunctionalCalculus.Unique
+
+/-!
+# Real powers defined via the continuous functional calculus
+
+This file defines real powers via the continuous functional calculus (CFC) and builds its API.
+This allows one to take real powers of matrices, operators, elements of a C⋆-algebra, etc. The
+square root is also defined via the non-unital CFC.
+
+## Main declarations
+
++ `CFC.nnrpow`: the `ℝ≥0` power function based on the non-unital CFC, i.e. `cfcₙ NNReal.rpow`
+  composed with `(↑) : ℝ≥0 → ℝ`.
++ `CFC.sqrt`: the square root function based on the non-unital CFC, i.e. `cfcₙ NNReal.sqrt`
++ `CFC.rpow`: the real power function based on the unital CFC, i.e. `cfc NNReal.rpow`
+
+## Implementation notes
+
+We define two separate versions `CFC.nnrpow` and `CFC.rpow` due to what happens at 0. Since
+`NNReal.rpow 0 0 = 1`, this means that this function does not map zero to zero when the exponent
+is zero, and hence `CFC.nnrpow a 0 = 0` whereas `CFC.rpow a 0 = 1`. Note that the non-unital version
+only makes sense for nonnegative exponents, and hence we define it such that the exponent is in
+`ℝ≥0`.
+
+## Notation
+
++ We define a `Pow A ℝ` instance for `CFC.rpow`, i.e `a ^ y` with `A` an operator and `y : ℝ` works
+  as expected. Likewise, we define a `Pow A ℝ≥0` instance for `CFC.nnrpow`.
+
+## TODO
+
++ Relate these to the log and exp functions
++ Lemmas about how these functions interact with commuting `a` and `b`.
++ Prove the order properties (operator monotonicity and concavity/convexity)
+-/
+
+open scoped NNReal
+
+namespace CFC
+
+section NonUnital
+
+variable {A : Type*} [PartialOrder A] [NonUnitalNormedRing A] [StarRing A] [StarOrderedRing A]
+  [TopologicalRing A] [Module ℝ≥0 A] [SMulCommClass ℝ≥0 A A] [IsScalarTower ℝ≥0 A A]
+  [CompleteSpace A] [NonUnitalContinuousFunctionalCalculus ℝ≥0 (fun (a : A) => 0 ≤ a)]
+
+/- ## `nnrpow` -/
+
+/-- Real powers of operators, based on the non-unital continuous functional calculus. -/
+noncomputable def nnrpow (a : A) (y : ℝ≥0) : A := cfcₙ (fun x : ℝ≥0 => x ^ (y : ℝ)) a
+
+/-- Enable `a ^ y` notation for `CFC.nnrpow`. This is a low-priority instance to make sure it does
+not take priority over other instances when they are available. -/
+noncomputable instance (priority := 50) : Pow A ℝ≥0 where
+  pow a y := nnrpow a y
+
+@[simp]
+lemma nnrpow_eq_pow {a : A} {y : ℝ≥0} : nnrpow a y = a ^ y := rfl
+
+@[simp]
+lemma nnrpow_nonneg {a : A} {x : ℝ≥0} : 0 ≤ a ^ x := cfcₙ_predicate _ a
+
+lemma nnrpow_def {a : A} {y : ℝ≥0} : a ^ y = cfcₙ (fun x : ℝ≥0 => x ^ (y : ℝ)) a := rfl
+
+lemma nnrpow_add {a : A} {x y : ℝ≥0} (hx : 0 < x) (hy : 0 < y) :
+    a ^ (x + y) = a ^ x * a ^ y := by
+  simp only [nnrpow_def]
+  rw [← cfcₙ_mul (fun z : ℝ≥0 => z ^ (x : ℝ)) (fun z : ℝ≥0 => z ^ (y : ℝ)) a]
+  congr
+  ext z
+  have : (x : ℝ) + (y : ℝ) ≠ 0 := by exact_mod_cast (ne_of_gt (add_pos hx hy))
+  simp [NNReal.rpow_add' _ this]
```

Comment:
This `rw` was *extremely* slow (thank goodness for incremental compilation!). I spent a short amount of time debugging, and about an hour coming up with a solution. The issue seems to be the following: `fun_prop (disch := aesop)` succeeded, but was very slow. Moreover, it was clear `aesop` was doing some sort of lifting because `fun_prop` alone failed. So I figured, I'll add an `@[fun_prop]` lemma of the form `∀ y : ℝ≥0, Continuous fun x : ℝ≥0 ↦ x ^ (y : ℝ)` because then it will find this and we'll be set. That *did* allow `fun_prop` alone to succeed, but it was still slow. However, `apply Continuous.continuousOn; fun_prop` was instant. This suggested to me that `fun_prop` was going down some rabbit hole involving the coercion because too much was unfolded already. So, my solution is this: we add
```
namespace NNReal

noncomputable
abbrev nnrpow (a : ℝ≥0) (b : ℝ≥0) : ℝ≥0 := a ^ (b : ℝ)

-- maybe this could even be marked `@[simp]`
lemma nnrpow_def (a b : ℝ≥0) : nnrpow a b = a ^ (b : ℝ) := rfl

@[fun_prop]
lemma continuous_nnrpow_const (y : ℝ≥0) : Continuous (nnrpow · y) :=
  continuous_rpow_const zero_le_coe

end NNReal
```
Then change the following def and lemma to:
```
noncomputable def nnrpow (a : A) (y : ℝ≥0) : A := cfcₙ (NNReal.nnrpow · y) a

lemma nnrpow_def {a : A} {y : ℝ≥0} : a ^ y = cfcₙ (NNReal.nnrpow · y) a := rfl
```
It is a bit interesting that `abbrev NNReal.nnrpow` is enough, but it seems to be. Indeed, this allows for the quite fast and nice proof of this lemma:
```suggestion
lemma nnrpow_add {a : A} {x y : ℝ≥0} (hx : 0 < x) (hy : 0 < y) :
    a ^ (x + y) = a ^ x * a ^ y := by
  simp only [nnrpow_def]
  rw [← cfcₙ_mul _ _ a]
  congr! 2 with z
  exact_mod_cast NNReal.rpow_add' z <| ne_of_gt (add_pos hx hy)
```
(Note: it is possible to increase the speed further by providing a few `simp` lemmas for `NNReal.nnrpow` that allow us to prove `NNReal.nnrpow 0 x = 0` faster, because at this point the slowest part of the proof is the `aesop` calls to show this in the auto-params for `cfcₙ_mul`, but the `fun_prop` calls are now nearly instantaneous.)

This suggestion requires adding `NNReal.nnrpow_def` to some of the `simp` calls elsewhere.